File: c++/2018/solutions/day02-1.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <fstream>

bool letterAmount(std::string_view s, int amt) {
	std::unordered_map<char, int> counts{};
	counts.reserve(26);
	
	for (char c : s) {
		++counts[c];
	}
	
	for (auto pair : counts) {
		if (pair.second == amt) return true;
	}
	return false;
}

int main() {
	std::ifstream input{"inputs/day02-input.txt"};
	if (!input.is_open()) std::cout << "file not open\n";
	
	std::string s{};
	int twoLetters{};
	int threeLetters{};
	while (std::getline(input, s)) {
		if (letterAmount(s, 2)) ++twoLetters;
		if (letterAmount(s, 3)) ++threeLetters;
	}
	
	std::cout << twoLetters * threeLetters << '\n';
}

File: ruby/2018/day02-1.rb
input = File.open('../inputs/2018/day02-input.txt', 'r')

alphabet = "abcdefghijklmnopqrstuvwxyz"
twoLetters = 0
threeLetters = 0
input.each do |id|
    count = {}
    alphabet.chars.each do |c|
        count[c] = id.count(c)
    end
    if (count.has_value?(2)) then twoLetters += 1 end
    if (count.has_value?(3)) then threeLetters += 1 end
end

puts twoLetters * threeLetters

input.close
